### ISSUE TITLE
Gevent performance regression Python 3.9 -> Python 3.11

### ISSUE THREAD

**Issue by davidkhess** [gevent/gevent#1928]
* gevent version: 22.10.2 via pip install
* Python version: 

Python 3.9.9 (main, Nov 18 2021, 07:23:37) 
[Clang 13.0.0 (clang-1300.0.29.3)] on darwin

* Operating System: 
 
ProductName:	macOS
ProductVersion:	12.6
BuildVersion:	21G115

### Description:

I expected gevent to perform comparably to asyncio on Python 3.9 and that gevent not have a performance regression on Python 3.11.

All details and reproduction notes are here:

https://dev.to/skywind3000/performance-asyncio-vs-gevent-vs-native-epoll-bnl

### What I've run:

I replicated these results on the system described at the top of the issue on Python 3.9 (primarily to see if asyncio really is that much faster). I have not replicated the performance regression on Python 3.11 but trust the results on that benchmark page after replicating the situation on Python 3.9.

All details for reproduction are here: 

https://dev.to/skywind3000/performance-asyncio-vs-gevent-vs-native-epoll-bnl

Bottom line - has gevent fallen behind asyncio in performance and will it / can it catch up?

**Comment by davidkhess** [comment]
In researching this on my own, I’ve grown to suspect that all of the C stack switching that greenlet’s implementation does may be the root issue here. Native Python coroutines don’t have that problem when switching between them since Python frames are in the heap and not on the C stack. It seems it’s pointer management instead of stack copying. 

Assuming that’s correct, rather than from an interop angle, does it make sense to attempt to implement gevent on top of native coroutines in order to get their performance benefits? (Also found this related issue: #641)

I’ve been experimenting with trying to implement a PoC pure Python version of the greenlet API on top of native coroutines and haven’t had much luck so far. I think the major impediment there is coroutines give you no way of yielding control without actual await / yield and your callstack needs to be coroutines all the way down to the suspension point.

**Comment by davidkhess** [comment]
So, I have done a bit of experimentation and research on this and thought I'd document it here for the next person who comes along.

I put together a proof of concept of reimplementing the greenlet module using raw generators (*not* asyncio). It actually turned out pretty good except for the Python limitation that a generator really only encompasses the top most frame of it (i.e. still executing code in the generator function itself). Meaning, you can only yield from the topmost Python frame and if you are any deeper framewise you cannot yield execution of the generator. That's the same restriction that causes async / await coroutines to have to be nested all the way down in order to properly yield back to the asyncio event loop (since coroutines are just specialized generators).

I started to fork CPython in order to experiment around with the idea of adding the ability to yield a coroutine from a Python frame below the topmost one and ran into the issue that immediately made it clear why it would be very difficult to accomplish, explains why coroutines work the way they do and explains why greenlet has to do what it does with stack switching.

Traditionally, the CPython implementation actually places a new C stackframe on the process stack every time control passes to a new function. Generators work despite this because the restriction on yielding keeps them within that same single C stackframe. Thus, suspending one doesn't require preserving the state of the C stack (tricky and difficult), only the state of the currently executing python frame and switching to a different one (simpler). The moment a generator calls a non-generator function, you now have an additional C stackframe that can't be easily switched out when you want to switch to a different generator or the main thread of execution. And thus Stackless Python and the greenlet module were created and continue to exist.

I mentioned "traditionally" above because I found this Python issue that just happened to land some changes in this regard in Python 3.11:

https://bugs.python.org/issue45256 - Remove the usage of the C stack in Python to Python calls

The original author of Stackless / greenlet Christian Tismer even chimes in on that effort. The whole situation becomes clear quickly. Generators work the way they do and have the limitations they do because it's really hard to do anything more in the current CPython implementation due to how C frames are managed. Smart people involved know this and the CPython implementation is starting to move in the direction of Stackless / greenlet which should eventually remove these kinds of restrictions on generators and thus make true green threads possible natively in CPython (without requiring coroutines all the way down).

But as that discussion makes clear, it's going to require a lot more work than just this one change before greenlet isn't needed any longer. There are still many other ways that the C frame stack management can cause problems for managing green threads of execution. And who knows, maybe this trend and the resulting awareness will lead to greenlet being incorporated into CPython natively and thus delivering native green threads sooner.

Bottom line for me, greenlet may end up permanently slower than asyncio because of CPython implementation limitations, but the programming model of true green threads (vs. coroutines all the way down and the complexity and syntactic issues that come with it) I think is worth it.

At least for now.

**Comment by davidkhess** [comment]
After the above, I got interested in green thread technology and took a dive into the implementation of greenlet and of PyPy's stacklet. I also came across an interesting project for green threading in straight C: http://libdill.org

What I discovered is there isn't a whole lot of code to green threading implementations – but it is extremely tricky code. There are also two major differences in how they get implemented. greenlet and stacklet both use "stack slicing" which involves copying each greenlets stack too and from the main process stack. 

However, libdill uses a dedicated stack for each green thread. The stack is allocated on the heap and has a readonly guard page next to it to catch stack overflow. This is supposedly a big performance win as you no longer have to copy stacks around when switching between green threads. It's just pointer fixups. Of course, the tradeoff is increased memory usage.

I have it on my todo list to run a C-level profiler on the greenlet test that started this issue because I really want to validate the assumption behind my thoughts here - that stack copying is the root of the problem.

But in the meantime, finally a request for the maintainers.... If stack copying does prove to be the root of the performance issue, can greenlet be modified to support both stack slicing and dedicated stacks? For a web application server, we'd likely want all of the worker greenlets running on their own stacks. But, for non-performance sensitive uses of greenlet, we'd likely want them running with stack slicing so they don't have to worry about running out of stack.

**Comment by navytux** [comment]
For the reference there is a project that provides greenlet-like API, but uses dedicated stack for each microthread: https://github.com/tbodt/greenstack, https://github.com/python-greenlet/greenlet/issues/113.

**Comment by navytux** [comment]
> I think the major impediment there is coroutines give you no way of yielding control without actual await / yield and your callstack needs to be coroutines all the way down to the suspension point.

For the reference 2: coroutines, even async/await ones, are just generators.
`async def` and `await` is just another syntax for `def` and `yield from`. So
coroutine being a generator - it can be interacted with and run step-by-step
via `.send`/`.throw` API that generators provide. Asyncio provides such runner in form of Task:

https://github.com/python/cpython/blob/b1a74a182d8762bda51838401ac92b6ebad9632a/Lib/asyncio/tasks.py#L78-L354

and it is quite doable to implement such a runner in a custom way - see e.g. here:

https://github.com/zopefoundation/ZEO/blob/e6de4796b98ce8b41bb5dd36ec381e4b248e2747/src/ZEO/asyncio/futures.py#L231-L392

This way a gevent loop could be also running coroutines spawned via asyncio by sending/receiving to/from them as needed.

**Comment by davidkhess** [comment]
If I'm understanding you correctly, I don't think that approach will work. I built a proof of concept of implementing greenbelts using generators and found that it won't work because an actual `yield` statement must be hit in a generator for it to be eligible for resumption via `.send()`. If you attempt to `.send()` to a coroutine already in a `.send()` you get an error from the interpreter saying it is not allowed.

(The reason why becomes clear once you realize the restriction that generators must be executing in their top level frame when they yield. Otherwise, the interpreter has some intervening C stack that needs to be swapped out which it has no capability to do.)

So, you must use `yield` to give up control in anything spawned via generators. And `yield` can only happen executed as a python statement and only from the top level frame of the generator itself. It can't be done in any deeper stack frames.

**Comment by davidkhess** [comment]
Reading your comment again, is your point that gevent could host coroutine-based Tasks along with greenlets? I certainly agree with that.

I'm interested in the opposite case. Can coroutines be used to implement and host greenlets? The answer appears to be no.

**Comment by navytux** [comment]
@davidkhess thanks for feedback. Yes, I've put a note regarding gevent loop hosting coroutines just for the reference. My main point was about greenstack.
